Docs: Add docs for ad-hoc filters in data source plugins (#33467)
grafana/grafana: docs/sources/developers/plugins/add-support-for-adhoc-filters.md

+++
title = "Ad-hoc filters for data source plugins"
+++

# Ad-hoc filters for data source plugins

> If your data source supports filtering, such as `WHERE` for SQL, consider adding support for [ad-hoc filters](TODO).

Reviewer: Do our built in SQL plugins support ad-hoc filters?  We were looking to add this to SAP Hana (SQL based) but some challenges arose:
* How to inject a filter into complex sql queries:  
  * an outer query along with subqueries with where clauses (where do we inject the filter?)
* How to get the tagKeys?  Seems this should be query based, but there is no way to pass a query (like you can with Query type variables) - this is a limitation we hit a lot
* getTagValues() doesn't accept a key/value pair (only the "text" value from the keys dropdown).  For SQL queries (and other filter apis) we often need an "id" and a display value.  So we can show a nice display value in the Select, but supply the associated id to the where clause.  Another limitation we hit a lot.

Author: Searching for getTagKeys tells me that Elasticsearch, InfluxDB, and Prometheus seems to be the only ones that use ad-hoc filters. Should we change the example perhaps?

## Add support for ad-hoc filters variables to your data source

To use your data source for populating ad-hoc filter variables, it needs to implement the `getTagKeys` and `getTagValues` methods from DataSourceApi.

`getTagKeys` returns an array of keys that the can be used to filter on.

```ts
async getTagKeys(): Promise<MetricFindValue[]> {
  return [
    { text: 'Organization' },
    { text: 'Project' },
  ];
}
```

> `getTagValues` is called whenever the user lists the values for a specific key. The `options` argument has a property called `key`, which corresponds to the `text` property of the selected tag key. You can use the `key` to determine the values that should be returned for that key.

Reviewer 2: That should be returned or that are returned?

```ts
async getTagValues(options: any): Promise<MetricFindValue[]> {
  switch (options.key) {
    case 'Organization':
      return [{ text: 'Grafana Labs' }];
    case 'Project':
      return [{ text: 'Grafana' }, { text: 'Loki' }, { text: 'Tempo' }];
    default:
      return [];
  }
}
```

## Use ad-hoc filters in your data source

> In contrast to the other variable types, you don't interpolate ad-hoc filter variables. Instead you need to list them using the `getVariables` method on `TemplateSrv`.

Reviewer 2: ```suggestion
Unlike other variable types, you don't interpolate ad-hoc filter variables. Instead, you list them using the `getVariables` method on `TemplateSrv`.
```

> > **Note:** Since ad-hoc filters have yet to be defined in the `VariableModel` interface, you need to annotate the variable as `any` to access the `filters` property.

Reviewer 2: ```suggestion
> **Note:** Because ad-hoc filters are not defined in the `VariableModel` interface, you must annotate the variable as `any` to access the `filters` property.
```

```ts
const adhocFilters = getTemplateSrv()

>   .getVariables()

Reviewer: I've always used: getTemplateSrv().getAdhocFilters("datasource name")

Author: Didn't know about this one! I'll check it out and compare

Author: Doesn't seem like getAdhocFilters is exposed by [TemplateSrv](https://grafana.com/docs/grafana/latest/packages_api/runtime/templatesrv/)?

Author: Hm. Seems like we have an internal and a public TemplateSrv, and that getAdhocFilters is only exposed by the internal one. Since this is the public docs, it feels like we should stick with getVariables for now? WDYT?

Reviewer: How do we import the public getTemplateSrv?  Maybe we should add that.

  .filter((variable) => variable.type === 'adhoc')
  .flatMap((variable: any) => variable.filters);
```

The `adhocFilters` looks something like this:

```ts
[
  {
    condition: "",
    key: "Organization",
    operator: "=",
    value: "Grafana Labs"
  },
  {
    condition: "",
    key: "Project",
    operator: "=",
    value: "Grafana"
  }
]
```

> The next step is to map the defined filters to the query. This depends on the query language that is used by your data source.

Reviewer 3: I feel like this is not enough. We need to be able to associate metadata with a key and value. For example, some datasources like dynatrace force querying on a key. `Grafana Labs` may actually just be a descriptive name, and the datasource requires querying by it's id or key which might look like `ORG-1A2BC`. I believe there are some other fields returned. We should look closer and ensure that we are not leaving this out of the documentation.

> The next step is to map the defined filters to the query. This depends on the query language that is used by your data source.

Reviewer 2: This seems like we are missing content. Where is the next step? Perhaps link to the topic where you learn how to do this? At the very least tell them how to find the resources they need.
